packages/core/src/types.ts

```diff
@@ -0,0 +1,370 @@
+
+	/**
+	 * Any parsing errors, whether fatal or not. (only when ok: false)
+	 */
+	errors?: Array<FileError>;
+
+	/**
+	 * Any additional data that the parser wants to provide.
+	 */
+	[key: string]: any;
+}
+
+export type ParseResult = OkParseResult | NotOkParseResult;
+
+/**
+ * Represents inline configuration found in the source code.
+ */
+interface InlineConfigElement {
+	/**
+	 * The location of the inline config element.
+	 */
+	loc: SourceLocation;
+
+	/**
+	 * The interpreted configuration from the inline config element.
+	 */
+	config: {
+		rules: RulesConfig;
+	};
+}
+
+/**
+ * Represents the basic interface for a source code object.
+ */
+interface SourceCodeBase {
+	/**
+	 * Root of the AST.
+	 */
+	ast: SyntaxElement;
+
+	/**
+	 * Traversal of AST.
+	 */
+	traverse(): Iterable<TraversalStep>;
+
+	/**
+	 * Applies configuration found inside of the source code. This method is only
+	 * called when ESLint is running with inline configuration allowed.
+	 */
+	applyInlineConfig?(): void;
+
+	/**
+	 * Applies language options passed in from the ESLint core.
+	 */
+	applyLanguageOptions?(): void;
+
+	/**
+	 * Return all of the inline areas where ESLint should be disabled/enabled
+	 * along with any problems found in evaluating the directives.
+	 */
+	getDisableDirectives?(): Array<{
+		directives: Array<Directive>;
+		problems: Array<FileProblem>;
+	}>;
+
+	/**
+	 * Returns an array of all inline configuration nodes found in the
+	 * source code.
+	 */
+	getInlineConfigNodes?(): Array<SyntaxElement>;
+
+	/**
+	 * Applies configuration found inside of the source code. This method is only
+	 * called when ESLint is running with inline configuration allowed.
+	 */
+	applyInlineConfig?(): Array<{
+		configs: Array<InlineConfigElement>;
+		problems: Array<FileProblem>;
+	}>;
+
+	/**
+	 * Called by ESLint core to indicate that it has finished providing
+	 * information. We now add in all the missing variables and ensure that
+	 * state-changing methods cannot be called by rules.
+	 * @returns {void}
+	 */
+	finalize?(): void;
+}
+
+/**
+ * Represents the source of a text file being linted.
+ */
+export interface TextSourceCode extends SourceCodeBase {
+	/**
+	 * The body of the file that you'd like rule developers to access.
+	 */
+	text: string;
+}
+
+/**
+ * Represents the source of a binary file being linted.
+ */
+export interface BinarySourceCode extends SourceCodeBase {
+	/**
+	 * The body of the file that you'd like rule developers to access.
+	 */
+	body: Uint8Array;
+}
+
+export type SourceCode = TextSourceCode | BinarySourceCode;
+
+/**
+ * Represents a traversal step visiting the AST.
+ */
+export interface VisitTraversalStep {
+	type: "visit";
+	kind: 1;
+	target: SyntaxElement;
+	phase: 1 /* enter */ | 2 /* exit */;
+	args: Array<any>;
+}
+
+/**
+ * Represents a traversal step calling a function.
+ */
+export interface CallTraversalStep {
+	type: "call";
+	kind: 2;
+	target: string;
+	phase?: string;
+	args: Array<any>;
+}
```

Comment:
Same here, there's no `kind` on these objects.

https://github.com/eslint/eslint/blob/e2b16e2b72606162dce3d804bc80186b6c5ec0f9/lib/languages/js/source-code/source-code.js#L1261-L1265

Also, I'm not sure if we should specify `phase` here since ESLint is not using it.